lib/actions/deploy.coffee

```diff
@@ -0,0 +1,168 @@
+getBuilderPushEndpoint = (baseUrl, owner, app) ->
+	"https://builder.#{baseUrl}/v1/push?owner=#{owner}&app=#{app}"
+
+parseInput = (params, options) ->
+	if not params.appName?
+		throw new Error('Need an application to deploy to!')
+	appName = params.appName
+	image = undefined
+	if params.image?
+		if (options.build or options.context?)
+			throw new Error('Build and context parameters are not applicable when specifying an image')
+		options.build = false
+		image = params.image
+	else if options.build
+		context = options.context || '.'
+	else
+		throw new Error('Need either an image or a build flag!')
+
+	return [appName, options.build, context, image]
+
+pushProgress = (imageSize, request, timeout = 250) ->
+	process.stdout.write('Initialising...')
+	progressReporter = setInterval ->
+		sent = request.req.connection._bytesDispatched
+		percent = (sent / imageSize) * 100
+		if percent >= 100
+			clearInterval(progressReporter)
+			percent = 100
+		process.stdout.clearLine()
+		process.stdout.cursorTo(0)
+		process.stdout.write("Uploaded #{percent.toFixed(1)}%")
+		console.log() if percent == 100
+	, timeout
+
+# A regex for parsing;
+# * A hostname e.g. registry.resin.io
+# * An appname, which consists of 1 or more alphanumeric chars
+# * A commit hash (40 alphanumeric characters)
+resinImageRegex = /.*\/(\w+\/)(\w{40})/
+
+updateFleet = (appName, image, token, url) ->
+	resin = require('resin-sdk-preconfigured')
+
+	if match = resinImageRegex.exec(image)
+		hash = match[2]
+	else
+		throw new Error('Could not parse reply from remote server. Perhaps you need to update resin-cli?')
+
+	resin.models.application.get(appName)
+	.then (app) ->
+		console.log("Deploying #{hash} to fleet...")
+		resin.pine.patch
+			resource: 'application'
+			id: app.id
+			body:
+				commit: hash
+
+getBundleInfo = (options) ->
+	Promise = require('bluebird')
+	helpers = require('../utils/helpers')
+
+	Promise.try(-> helpers.getAppInfo(options.appName))
+	.then (app) ->
+		[app.arch, app.device_type]
+
+module.exports =
+	signature: 'deploy <appName> [image]'
+	description: 'Deploy a container to a resin.io application'
+	help: '''
+		Use this command to deploy and optionally build an image to an application.
+
+		Usage: deploy <appName> ([image] | -b [-c build-dir])
+
+		Note: If building with this command, all options supported by `resin build`
+		are also support with this command.
+
+		Examples:
+		$ resin deploy myApp -b -c myBuildDir/
+		$ resin deploy myApp myApp/myImage
+	'''
+	permission: 'user'
+	options: [
+		{
+			signature: 'build'
+			boolean: true
+			description: 'Build image then deploy'
+			alias: 'b'
+		},
+		{
+			signature: 'context'
+			parameter: 'context'
+			description: 'The context to use when building the image'
+			alias: 'c'
+		}
+	]
+	action: (params, options, done) ->
+		Promise = require('bluebird')
+		request = require('request')
+		tmp = Promise.promisifyAll(require('tmp'))
```

Comment:
I tend not to use `promisifyAll` when we only need a method or two, you can check the usage of `tmp` in other modules here